third_party/move/move-compiler-v2/src/pipeline/reference_safety_processor.rs

```diff
@@ -1468,14 +1476,124 @@ impl<'env, 'state> LifetimeAnalysisStep<'env, 'state> {
+    /// semantics we may want to investigate to relax them, because they appear unnecessary
+    /// strict).
+    ///
+    /// When a reference is frozen, there must not exist any other mutable reference
+    /// pointing to the same location. It is, however, ok if the currently
+    /// frozen reference is used later again. This seems to be over-restrictive
+    /// since it shouldn't matter whether copies of the mutable reference exist
+    /// as long as they aren't passed at the same time as an argument to a function,
+    /// i.e. as long as they do not create aliasing.
+    ///
+    /// The above condition can be violated in two situations:
+    ///
+    /// a. The frozen reference has siblings which also mutably borrow the same parent.
+    /// b. There exists a mutable reference, alive after this program point, which
+    ///    borrows the same location but is not derived from the reference we are
+    ///    freezing.
+    fn freeze_ref(
+        &mut self,
+        code_offset: CodeOffset,
+        explicit: bool,
+        dest: TempIndex,
+        src: TempIndex,
+    ) {
         let label = *self.state.label_for_temp(src).expect("label for reference");
         let target = self.state.replace_ref(dest, code_offset, 0);
         self.state.add_edge(label, BorrowEdge {
             kind: BorrowEdgeKind::Freeze,
             loc: self.cur_loc(),
             target,
-        })
+        });
+        if let Some(label) = self.state.label_for_temp(src) {
+            // Handle case (a): search for any siblings which mutably borrow the same
+            // parent.
+            let qualifier = if explicit { "" } else { "implicitly " };
+            for (parent, edge) in self.state.parent_edges(label) {
+                for sibling_edge in self.state.children(&parent) {
+                    if &sibling_edge.target == label {
+                        continue;
+                    }
+                    if self.state.is_mut_path(sibling_edge) {
+                        self.error_with_hints(
+                            self.cur_loc(),
+                            format!(
+                                "cannot {}freeze {}  since multiple mutable references exist",
+                                qualifier,
+                                self.display(src)
+                            ),
+                            format!("{}frozen here", qualifier),
+                            vec![
+                                self.borrow_edge_info("originating ", edge),
+                                self.borrow_edge_info("conflicting ", sibling_edge),
+                            ]
+                            .into_iter(),
+                        )
+                    }
+                }
+            }
+            // Handle case (b): check whether there is any alive mutable reference
+            // which overlaps with the frozen reference.
+            for (temp, other_label) in self.state.temp_to_label_map.iter() {
```

Comment:
Could we directly use `leaves.get(&label)` to get all temps first so that there is not need to iterate over this for loop?

Reply:
AFAICT that is not working. We need to walk over all labels and if they are leaves do one thing else the other thing. Also this is really a DFS search, because we dynamically need to extend `todo`.

Reply:
Yeah, we need this iteration but we could iterate like`for temp in self.state.leaves().get(&label)` so that we don't need to check `if other_label == label` below? Also the condition `!self.ty(*temp).is_mutable_reference()` may not be needed because they must be mutable references?

Reply:
- We need the outer loop since `temp_to_label_map` is TempIndex -> Label, and not Label -> TempIndex. We basically need to iterate because we need the reverse of this map. So via that iteration, we essentially establish TempIndex -> Set(Label)
- We not only interested in parents but also parents of leaves, thats why leaves().get(label) isn't really that helpful.
- The map can contain both mutable and immutable references, therefore the test is needed.

Maybe I'm not seeing what you mean.

Reply:
It seems that I misunderstood the meaning of `self.leaves().get(label)`. I was just thinking if `other_label` needs to be equal to `label`, why not just return all temps corresponding to label.